File: crates/broadcast/src/signaling/messages.rs
use mediasoup::rtp_parameters::{MediaKind, RtpCapabilities, RtpCapabilitiesFinalized, RtpParameters};
use mediasoup::data_structures::{DtlsParameters, IceCandidate, IceParameters};
use mediasoup::transport::TransportId;
use mediasoup::consumer::ConsumerId;
use mediasoup::producer::ProducerId;

use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransportOptions {
    pub id: TransportId,
    pub dtls_parameters: DtlsParameters,
    pub ice_candidates: Vec<IceCandidate>,
    pub ice_parameters: IceParameters,
}

#[derive(Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerboundMessage {
    Begin {
        channel_id: String
    },
    Init {
        rtp_capabilities: RtpCapabilities
    },
    Connect {
        dtls_parameters: DtlsParameters
    },
    Consume,
    Resume {
        id: ConsumerId
    }
}

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Consume {
    pub id: ConsumerId,
    pub producer_id: ProducerId,
    pub kind: MediaKind,
    pub rtp_parameters: RtpParameters
}

#[derive(Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ClientboundMessage {
    Init {
        producers: Vec<ProducerId>,
        transport: TransportOptions,
        router_rtp_capabilities: RtpCapabilitiesFinalized
    },
    Connected,
    #[serde(rename_all = "camelCase")]
    Consuming {
        consume: Vec<Consume>
    }
}

File: packages/hyperspeed.js/src/messages.ts
import type { MediaKind, RtpCapabilities, RtpParameters } from "mediasoup-client/lib/RtpParameters";
import type { DtlsParameters, IceCandidate, IceParameters } from "mediasoup-client/lib/Transport";

export interface TransportOptions {
    id: string;
    dtlsParameters: DtlsParameters;
    iceCandidates: IceCandidate[];
    iceParameters: IceParameters
}

export interface Consume {
    id: string;
    producerId: string;
    kind: MediaKind;
    rtpParameters: RtpParameters;
}

export type ServerboundMessage = (
    {
        type: 'Begin',
        channel_id: string
    } | {
        type: 'Init',
        rtp_capabilities: RtpCapabilities
    } | {
        type: 'Connect',
        dtls_parameters: DtlsParameters,
    } | {
        type: 'Consume'
    } | {
        type: 'Resume',
        id: string
    }
)

export type ClientboundMessage = (
    {
        type: 'Init',
        producers: string[],
        transport: TransportOptions,
        router_rtp_capabilities: Required<RtpCapabilities>
    } | {
        type: 'Connected'
    } | {
        type: 'Consuming',
        consume: Consume[]
    }
)
